README.md
# Connect 4: Minimax and Alpha-Beta Pruning

[![Hits](https://hits.sh/github.com/vmsaif/connect4-with-minimax-algorithm-in-java.svg?label=Visits&color=100b75)](https://hits.sh/github.com/vmsaif/connect4-with-minimax-algorithm-in-java/)

## Logic and Design of Program
The program is a digital version of the classic Connect 4 game. The game is played on a 6x7 grid, where two players, the user and an AI will take turns dropping white and black discs respectively into the grid. The goal of the game is to connect four discs of the same color vertically, horizontally, or diagonally. The AI uses the minimax algorithm with alpha-beta pruning to determine the best move. 

When the game starts, the user is presented with a blank grid. The user can click on a column to drop a disc into that column. The program checks if the move is valid (i.e., the column is not full), and then updates the grid with the new disc. The program then checks if the move resulted in a win or a draw. If neither, the program switches to the other player's turn.

### Minimax and Alpha-Beta Pruning
The AI uses the minimax algorithm with alpha-beta pruning to determine the best move. The AI will look ahead in depth of 4 moves recursively. The AI will then choose the move that results in the highest score. The score will prioritize winning moves in fewer moves, and losing moves in more moves.

The AI will also use alpha-beta pruning to reduce the number of nodes that need to be evaluated. Alpha-beta pruning will reduce the number of nodes that are bad moves by either the AI or the user. This will make the AI more efficient and faster. 

## About This Repository

This repository uses CI/CD to automatically build and release the program for Windows, macOS, and Linux. The program is packaged as a JAR file. The JAR file is then packaged into an EXE file for Windows, a DMG file for macOS, and a DEB file for Linux. The program is built using GitHub Actions and is released using GitHub Releases. The program is built for the following platforms:

- Windows (64-bit)
- macOS (64-bit)
- Linux (64-bit)
- JAR File

You can find them here in the [Releases](https://github.com/vmsaif/connect4-with-minimax-algorithm-in-java/releases) section.

## Running the Program

- Make sure you have JRE installed on your computer. You can download JRE [here](https://adoptium.net/) or [here](https://www.oracle.com/java/technologies/javase-jre8-downloads.html).

Downloading the jar file is preferred. You can download the release version of the Game from the latest release [here](https://github.com/vmsaif/connect4-with-minimax-algorithm-in-java/releases). The program is packaged as a JAR file. You can run the program by double-clicking on the JAR file.

## **Ubuntu**

### Option 1: Downloading the DEB File (simpler)

1. Open the terminal and navigate to the directory where you downloaded the .deb file.

```bash
cd ~/Downloads
```

2. Run the following command to install the package:

```bash
sudo dpkg -i ./connect4-v1.1.0-linux_amd64.deb # Replace the filename/version with the name of the DEB file you downloaded
```
3. If you get an error, run the following command to install the dependencies:

```bash
sudo apt-get install -f
```

4. Run the following command to start the game:

You can start the game by searching for it in the applications menu named `connect4` or by running the following command in the terminal:
```bash
/opt/connect4/bin/connect4
```


### Option 2: Downloading the JAR File

1. **Installing Java** (if it's not already installed):
   ```bash
   sudo apt update
   sudo apt install default-jre
   ```

2. **Navigate to the Download Location**:
   ```bash
   cd ~/Downloads
   ```

3. **Provide Execute Permissions** (Optional, but useful if you want to execute it directly):
   ```bash
   chmod +x connect4-1.0.0.jar # Replace the filename with the name of the JAR file you downloaded
   ```

4. **Run the JAR File**:
   ```bash
   java -jar connect4-1.0.0.jar # Replace the filename with the name of the JAR file you downloaded
   ```

## **macOS**
There are 2 methods. Downloading the dmg file or downloading the jar file. 
### Option 1: Downloading the DMG File (simpler)
I am not a registered Apple developer, That is why you will get a warning when you try to open the app. Thus, you will have to bypass the security settings to run the app.
Here are the steps:

1. Click on the .dmg file you have downloaded. 

2. Drag the app to the Applications folder.

3. In the Finder <img src="https://help.apple.com/assets/63BCA927AAE78C58DD7FBE35/63BCA92EAAE78C58DD7FBE76/en_US/058e4af8e726290f491044219d2eee73.png" alt="" height="15" width="15" originalimagename="SharedGlobalArt/AppIconTopic_Finder.png"> on your Mac, locate the Connect4 app.

   (Don’t use Launchpad to do this. Launchpad doesn’t allow you to access the shortcut menu.)

4. Press and hold Control then click the app icon.

5. Click Open.

   The app is saved as an exception to your security settings, and you can open it in the future by double-clicking it just as you can any registered app.


### Option 2: Downloading the JAR File
Make sure you have JRE installed on your computer. You can download JRE [here](https://adoptium.net/) or [here](https://www.oracle.com/java/technologies/javase-jre8-downloads.html).

1. **Navigate to the Download Location** (using Terminal):
   ```bash
   cd ~/Downloads
   ```

3. **Provide Execute Permissions** (Optional):
   ```bash
   chmod +x connect4-1.0.0.jar # Replace the filename with the name of the JAR file you downloaded
   ```

4. **Run the JAR File**:
   ```bash
   java -jar connect4-1.0.0.jar # Replace the filename with the name of the JAR file you downloaded
   ```

## **Windows**

### Option 1: Downloading the EXE File (simpler)

1. Open the folder where you downloaded the EXE file.
2. Double-click on the EXE file to run it.

At this point, windows smart screen will block the app from running. As I am not a signed developer by Microsoft, this warning will appear. 

3. If you get the warning, click on "More Info" and then click on "Run Anyway".
4. Wait for the installation to complete.
5. Run the game from the Desktop shortcut.

### Option 2: Downloading the JAR File

1. **Confirming Java**:

   Make sure you have JRE installed on your computer. You can download JRE [here](https://adoptium.net/) or [here](https://www.oracle.com/java/technologies/javase-jre8-downloads.html).

   Ensure that Java is added to your `PATH` so that it can be accessed from the command prompt. During the jre installation, You will have option to add to your path. Is is unchecked by default. Make sure to enable it.  To check if Java is installed, run the following command in Command Prompt or PowerShell:
   ```bash
   java -version
   ```
   If Java is installed, you should see the version number. If you see an error, you need to install Java. See above for instructions on installing Java.

2. **Navigate to the Download Location** (using Command Prompt or PowerShell):
   ```bash
   cd c:\Users\username\Downloads # Replace username with your username
   ```

3. **Run the JAR File**:
   ```bash
   java -jar connect4-1.0.0.jar
   ```

In all cases, once the commands are followed, the Java application packaged inside the JAR file should start running.


## Compiling and Running (Advanced)
### Compiling
The game is written in Java and can be compiled using the following steps:

- Go to the root directory
- Run the command `javac -d bin src/*.java` to compile the Java file.
- Run the command `java -cp bin/ App` to run the game.


## Bugs:
I have not found any bugs in the program yet. If you find any, please let me know by opening an issue from [here](https://github.com/vmsaif/connect4-with-minimax-algorithm-in-java/issues/)

## Acknowledgements
- [Icon](https://icon-icons.com/icon/grid-board-cell-table/187134)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

## [v1.1.0] - 2023-09-03

### Added

- added reset button to reset the game

## [v1.0.0] - 2023-09-03

### Added

- Initial release.


src/Circle.java
/* -----------------------------------------------------------------------------
    Author: Saif Mahmud
    Date: 2023-02-08 (yyyy-dd-mm)

    Description: This class is creates and draws a circle icon.
*/
import java.awt.Color;
import java.awt.Component;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;

import javax.swing.Icon;

public class Circle implements Icon {
    private Color fillColor;
    private Color borderColor;
    private int size;

    public Circle(Color fillColor, Color borderColor) {
        this.fillColor = fillColor;
        this.borderColor = borderColor;
        this.size = 40;
    }

    @Override
    public void paintIcon(Component c, Graphics g, int x, int y) {
        Graphics2D g2d = (Graphics2D) g.create();
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON); // Anti-alias! 
        
        // Draw the circle
        g2d.setColor(fillColor);
        g2d.fillOval(x, y, size, size);
        
        // Draw the border
        g2d.setColor(borderColor);
        g2d.drawOval(x, y, size, size);
        
        g2d.dispose(); // Clean up
    }

    public int getIconWidth() {
        return size;
    }

    public int getIconHeight() {
        return size;
    }
}

src/Computer.java
/* -----------------------------------------------------------------------------
    Author: Saif Mahmud
    Date: 2023-02-08 (yyyy-dd-mm)

    Description: This class is the computer player. It uses the minimax algorithm and alpha-beta pruning to determine the best move.
*/

public class Computer {

    private char[][] board;
    private int MAX_DEPTH;

    private char EMPTY;
    private char WHITE;
    private char BLACK;
    private int bestMove[]; // [0] = row, [1] = col
    private int winningLength;

    public Computer(char[][] board, int MAX_DEPTH, int WINNING_LENGTH) {
        this.winningLength = WINNING_LENGTH;
        this.board = board;
        this.MAX_DEPTH = MAX_DEPTH;
        this.EMPTY = Connect4.getEmpty();
        this.WHITE = Connect4.getWhite();
        this.BLACK = Connect4.getBlack();
        this.bestMove = new int[2];
    }


    // make the line buildup
    // check linebuildup for WHITE player
    public void makeComputerMove() {

        if(Connect4.isWhiteTurn() == false) {
            // computer's turn
            int alpha = Integer.MIN_VALUE;
            int beta = Integer.MAX_VALUE;            
            int evaluation = Integer.MIN_VALUE;
            
            for(int row = board.length-1; row >= 0 ; row--) {
                for(int col = 0; col < board[row].length; col++) {
                    if(board[row][col] == EMPTY) {
                        // check if this is a valid move
                        if(row == board.length-1 || board[row+1][col] != EMPTY) {
                            // try this move
                            board[row][col] = BLACK;
                            int temp = minimax(MAX_DEPTH, alpha, beta, false);

                            // now reset the move
                            board[row][col] = EMPTY; 

                            if(temp > evaluation) {
                                // got a better move
                                evaluation = temp;
                                bestMove[0] = row;
                                bestMove[1] = col;
                            }
                        }
                    }
                }
            } // end for loop
            // now make the best move
            board[bestMove[0]][bestMove[1]] = BLACK;
            Connect4.printBoard("BLACK");
            // now switch turns
            Connect4.switchTurn();
        }
    }

    // evaluate the board for a given player - this method is responsible for the heuristic of AI
    private int evaluateBoardForPlayer(char player) {
        int score = 0;

        // For each column, row, and direction
        for (int row = 0; row < board.length; row++) {
            for (int col = 0; col < board[0].length; col++) {
                for (int[] dir : new int[][]{{0,1}, {1,0}, {1,1}, {1,-1}}) {  // Right, Down, Down-right, Down-left
                    int countPlayerPieces = 0;
                    int countEmpty = 0;

                    // Check each cell in the line upto winningLength
                    for (int i = 0; i < winningLength; i++) {
                        int newRow = row + dir[0] * i;
                        int newCol = col + dir[1] * i;

                        // Check if the cell is on the board
                        if (newRow >= 0 && newRow < board.length && newCol >= 0 && newCol < board[0].length) {
                            if (board[newRow][newCol] == player) {
                                countPlayerPieces++;
                            } else if (board[newRow][newCol] == EMPTY) {
                                countEmpty++;
                            }
                        }
                    }

                    if (countPlayerPieces + countEmpty == winningLength) {
                        int initialScore = (winningLength * 10) + 10; // for winningLength = 4, initialScore = 50
                        for (int i = 1; i < winningLength; i++){
                            if (countPlayerPieces == winningLength - i) {
                                score += initialScore/i;  // Adjust as needed, I am just splitting by I to get 50, then 20, then 10 by this and next line.... For a winning length of 4
                                initialScore = initialScore - 10;
                            }
                        }
                    }
                }
            }
        }
        return score;
    }

    
    // minimax algorithm with alpha-beta pruning - handles the decision tree
    private int minimax(int depth, int alpha, int beta, boolean isWhitesTurn) {
        int result = 0;
        boolean stopSearch = false;
        boolean blackWon = Connect4.checkWin(BLACK);
        boolean whiteWon = Connect4.checkWin(WHITE);
        boolean draw = Connect4.checkDraw();

        // base case
        if(depth == 0 || blackWon || whiteWon || draw) {
            // evaluate the board from the computer's perspective
            if(blackWon){
                result = 1000 + depth;
            } else if(whiteWon){
                result = -(1000 + depth);
            } else if(draw){
                result = 0;
            } else {
                int blackScore = evaluateBoardForPlayer(BLACK);
                int whiteScore = evaluateBoardForPlayer(WHITE);
                result = blackScore - whiteScore;
            }
        } else { // recursive case
            if(!isWhitesTurn) {
                // minimizing player COMPUTER (BLACK)
                result = Integer.MAX_VALUE;
                for(int row = board.length-1; row >= 0 && !stopSearch; row--) {
                    for(int col = 0; col < board[row].length && !stopSearch; col++) {
                        if(board[row][col] == EMPTY) {

                            // check if this is a valid move
                            if(row == board.length-1 || board[row+1][col] != EMPTY) {
                                // try this move
                                board[row][col] = WHITE;
                                int temp = minimax(depth-1, alpha, beta, true);
                                // now reset the move
                                board[row][col] = EMPTY;
                                result = Math.min(result, temp);
                                beta = Math.min(beta, result);

                                // if poor beta value, stop searching
                                if(beta <= alpha) {
                                    stopSearch = true;
                                }
                            }
                        }
                    }
                } // end for loop
            } else {
                // maximizing player USER (WHITE)
                result = Integer.MIN_VALUE;
                for(int row = board.length-1; row >= 0 && !stopSearch; row--) {
                    for(int col = 0; col < board[row].length && !stopSearch; col++) {
                        if(board[row][col] == EMPTY) {
                            // check if this is a valid move
                            if(row == board.length-1 || board[row+1][col] != EMPTY) {
                                // try this move
                                board[row][col] = BLACK;
                                int temp = minimax(depth-1, alpha, beta, false);
                                // now reset the move
                                board[row][col] = EMPTY;
                                result = Math.max(result, temp);
                                alpha = Math.max(alpha, result);

                                // if beta is less, prune
                                if(beta <= alpha) {
                                    stopSearch = true; 
                                }
                            }
                        }
                    }
                }
            } // end for loop
        }
        
        return result;
    }

}// end class Computer



src/Connect4.java
/* -----------------------------------------------------------------------------
    Author: Saif Mahmud
    Date: 2023-02-08 (yyyy-dd-mm)
    
    Description: This program is a game of Connect 4. The user plays against the computer. 
    Computer uses the minimax algorithm and alpha-beta pruning to determine the best move.
*/
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

public class Connect4 extends JFrame {
    private static final int ROWS = 6;
    private static final int COLUMNS = 7;
    private static final int WINNING_LENGTH = 4;
    private static final int MAX_DEPTH = 4;

    private static final char EMPTY = ' ';
    private static final char WHITE = 'W';
    private static final char BLACK = 'B';

    private JPanel boardPanel;
    private JButton[] columnButtons;

    private static char[][] board = new char[ROWS][COLUMNS];
    private static boolean isWhiteTurn = true;
    private boolean gameOver = false;
    private Computer computer;

    public Connect4 (){}


    public void runGame() {
        initializeBoard();
        computer = new Computer(board, MAX_DEPTH, WINNING_LENGTH);
        setTitle("Connect 4");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        boardPanel = new JPanel(new GridLayout(ROWS + 1, COLUMNS));
        columnButtons = new JButton[COLUMNS];

        createDropButtons();
        addResetButton();

        // set each tile as an unclickable button with a white background to just add circles on top
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLUMNS; col++) {
                JButton button = new JButton();
                button.setEnabled(false);
                button.setPreferredSize(new Dimension(50, 50));
                button.setBackground(Color.WHITE);
                boardPanel.add(button);
            }
        }

        add(boardPanel);
        pack();
        setVisible(true);
    }

    private void createDropButtons() {
        for (int col = 0; col < COLUMNS; col++) {
            JButton button = new JButton("Drop");
            button.setActionCommand(Integer.toString(col));
            button.addActionListener(new ColumnButtonListener());
            columnButtons[col] = button;
            boardPanel.add(button);
        }
    }

    // make the moves when the user clicks on a drop button
    private class ColumnButtonListener implements ActionListener {
        public void actionPerformed(ActionEvent e) {
            if (!gameOver) {
                int column = Integer.parseInt(e.getActionCommand());
                makeUserMove(column);
                computer.makeComputerMove();
                
                setGridIcon();

                if (checkWin(WHITE)) {
                    JOptionPane.showMessageDialog(Connect4.this, "WHITE wins!");
                    gameOver = true;
                    
                } else if (checkWin(BLACK)) {
                    JOptionPane.showMessageDialog(Connect4.this, "BLACK wins!");
                    gameOver = true;
                    
                } else if (checkDraw()) {
                    JOptionPane.showMessageDialog(Connect4.this, "Draw!");
                    gameOver = true;
                    
                }
            }
        }
    }

private void addResetButton() {
    JPanel resetPanel = new JPanel();
    JButton resetButton = new JButton("Reset");
    resetButton.addActionListener(new ActionListener() {
        public void actionPerformed(ActionEvent e) {
            // Reset the game
            
            isWhiteTurn = true;
            gameOver = false;
            runGame();
        }
    });
    resetPanel.add(resetButton);
    add(resetPanel, BorderLayout.SOUTH);
    pack();
    setVisible(true);
}


    public void setGridIcon() {
        Component[] components = boardPanel.getComponents();
        int index = COLUMNS; // start at the bottom left of the board by reducing the indexes of drop buttons
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLUMNS; col++) {
                JButton button = (JButton) components[index];
                
                if (board[row][col] == WHITE) {
                    button.setIcon(new Circle(Color.WHITE, Color.BLACK));
                } else if (board[row][col] == BLACK) {
                    button.setIcon(new Circle(Color.BLACK, Color.BLACK));
                }
                index++;
            }
        }
    }


    public static void printBoard(String who) {
        System.out.println(who + " made a move.");
        System.out.println("    [0][1][2][3][4][5][6]");
        for (int row = 0; row < ROWS; row++) {
            System.out.print("["+row+"]");
            System.out.print("|");
            for (int col = 0; col < COLUMNS; col++) {
                System.out.print(board[row][col] + " |");
            }
            System.out.println();
        }
        System.out.println("---------------");
    }

    public void makeUserMove(int column) {
        boolean moveMade = false;
        for (int row = ROWS - 1; row >= 0 && !moveMade; row--) {
            if (board[row][column] == EMPTY) {
                if(isWhiteTurn) {
                    board[row][column] = WHITE;
                    moveMade = true;
                    printBoard("WHITE");
                } 
            }
        }

        // switch turns
        if(moveMade){
            switchTurn();
        }
    }

    // Switch turns
    public static void switchTurn() {
        isWhiteTurn = !isWhiteTurn;
    }

    public static boolean isWhiteTurn() {
        return isWhiteTurn;
    }

    public void initializeBoard() {
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLUMNS; col++) {
                board[row][col] = EMPTY;
            }
        }
    }

    public static boolean checkWin(char player) {
        
        // Check rows
        boolean result = false;

        for (int row = 0; row < ROWS && !result; row++) {
            for (int col = 0; col <= COLUMNS - WINNING_LENGTH && !result; col++) {
                boolean win = true;
                for (int i = 0; i < WINNING_LENGTH && win; i++) {
                    if (board[row][col + i] != player) {
                        win = false;
                    }
                }
                if (win) {
                    result = true;
                }
            }
        }

        // Check columns
        for (int col = 0; col < COLUMNS && !result; col++) {
            for (int row = 0; row <= ROWS - WINNING_LENGTH && !result; row++) {
                boolean win = true;
                for (int i = 0; i < WINNING_LENGTH && win; i++) {
                    if (board[row + i][col] != player) {
                        win = false;
                    }
                }
                if (win) {
                    result = true;
                }
            }
        }

        // Check diagonals
        for (int row = 0; row <= ROWS - WINNING_LENGTH && !result; row++) {
            for (int col = 0; col <= COLUMNS - WINNING_LENGTH && !result; col++) {
                boolean win = true;
                for (int i = 0; i < WINNING_LENGTH && win; i++) {
                    if (board[row + i][col + i] != player) {
                        win = false;
                    }
                }
                if (win) {
                    result = true;
                }
            }
        }

        for (int row = 0; row <= ROWS - WINNING_LENGTH && !result; row++) {
            for (int col = WINNING_LENGTH - 1; col < COLUMNS && !result; col++) {
                boolean win = true;
                for (int i = 0; i < WINNING_LENGTH && win; i++) {
                    if (board[row + i][col - i] != player) {
                        win = false;
                    }
                }
                if (win) {
                    result = true;
                }
            }
        }

        return result;
    }

    // Check if the board is full
    // return true if it is a draw
    public static boolean checkDraw() {
        boolean result = true;
        for(int row = 0; row < ROWS && result == true; row++) {
            for(int col = 0; col < COLUMNS && result == true; col++) {
                if(board[row][col] == EMPTY) {
                    result = false;
                }
            }
        }
        return result;
    }

    public static char getBlack() {
        return BLACK;
    }

    public static char getWhite() {
        return WHITE;
    }

    public static char getEmpty() {
        return EMPTY;
    }
} // end class Connect4